src/webgldrawer.js

```diff
@@ -237,169 +257,188 @@
             //iterate over tiled images and draw each one using a two-pass rendering pipeline if needed
             tiledImages.forEach( (tiledImage, tiledImageIndex) => {
 
-                let tilesToDraw = tiledImage.getTilesToDraw();
-
-                if(tilesToDraw.length === 0 || tiledImage.getOpacity() === 0){
-                    return;
-                }
-                let firstTile = tilesToDraw[0];
-
-                let useContext2dPipeline = ( tiledImage.compositeOperation ||
-                    this.viewer.compositeOperation ||
-                    tiledImage._clip ||
-                    tiledImage._croppingPolygons ||
-                    tiledImage.debugMode
-                );
-
-                let useTwoPassRendering = useContext2dPipeline || (tiledImage.opacity < 1) || firstTile.hasTransparency;
-
-                // using the context2d pipeline requires a clean rendering (back) buffer to start
-                if(useContext2dPipeline){
-                    // if the rendering buffer has image data currently, write it to the output canvas now and clear it
-
+                if(tiledImage.isTainted()){
+                    // first, draw any data left in the rendering buffer onto the output canvas
                     if(renderingBufferHasImageData){
                         this._outputContext.drawImage(this._renderingCanvas, 0, 0);
+                        // clear the buffer
+                        gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+                        gl.clear(gl.COLOR_BUFFER_BIT); // clear the back buffer
+                        renderingBufferHasImageData = false;
                     }
 
-                    // clear the buffer
-                    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
-                    gl.clear(gl.COLOR_BUFFER_BIT); // clear the back buffer
-                }
-
-                // First rendering pass: compose tiles that make up this tiledImage
-                gl.useProgram(this._firstPass.shaderProgram);
+                    // next, use the backup canvas drawer to draw this tainted image
+                    const canvasDrawer = this._getBackupCanvasDrawer();
```

Comment:
For applications that have potentially mixed tile sources, I want to be able to use the "best" drawer (i.e. the one specified when creating the viewer") for any `TiledImage`s that it the drawer is able to draw, and only fall back when necessary, rather than downgrading all images by replacing the entire drawer.

If a developer wants to, they could listen for the to-be-implemented `drawer-failure` event (or whatever we name it) and replace the whole drawer themselves.

Reply:
My point was:
- it is currently not easy to replace drawer, you would have to probably recreate the whole OSD instance so having implemented this fallback would help users too, supposing OSD re-instantion is not enough
- in most cases, if one of your data sources is tained, all of them are
- vanila WebGL drawing is less performant than canvas for obivous reasons, so that when you use webgl renderer, you
   - either want to get rid of seams -> then you get the seams with faulty data anyway
   - want to employ custom GPU processing features -> then the fallback per item image does not work anyway

So to me it seems like unnecessary feature, and supporting general drawer fallback replacement on the global level is a) cleaner and b) more 'stable' solution...

Reply:
> * it is currently not easy to replace drawer,

It actually shouldn't be too hard to swap out the drawer. Maybe we should add an API to do so.

> * in most cases, if one of your data sources is tained, all of them are

There are plenty of use-cases where an OSD viewer can be used to view images from across the web, which are not all under the control of the front-end developer. I think it would be a mistake to totally change the viewer away from webgl as soon as a single image is opened that doesn't have CORS configured.

> So to me it seems like unnecessary feature, and supporting general drawer fallback replacement on the global level is a) cleaner and b) more 'stable' solution...

I disagree, for the above reasons. The fact that the `CanvasDrawer` itself is used as the fallback (rather than copy/pasting code from that drawer) makes this approach pretty robust.